File: python/Solution841.py
class Solution:
    rooms_status = []

    def canVisitAllRooms(self, rooms):
        global rooms_status
        nums = len(rooms)
        rooms_status = [False for _ in range(nums)]
        rooms_status[0] = True
        self.dfs(rooms, 0)
        return False not in rooms_status

    def dfs(self, rooms, i):
        global rooms_status
        key_list = rooms[i]
        for key in key_list:
            if (not rooms_status[key]):
                rooms_status[key] = True
                self.dfs(rooms, key)


if __name__ == '__main__':
    result = Solution().canVisitAllRooms([[6, 7, 8], [5, 4, 9], [], [8], [4], [], [1, 9, 2, 3], [7], [6, 5], [2, 3, 1]])
    print(result)

File: java/src/Solution841.java
import java.util.List;

public class Solution841 {
    boolean[] visited;
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        visited = new boolean[rooms.size()];
        visited[0] = true;
        dfs(rooms, 0);

        // 判断
        for (boolean v : visited) {
            if (!v) {
                return false;
            }
        }
        return true;
    }

    public void dfs(List<List<Integer>> rooms, int key) {
        List<Integer> key_list = rooms.get(key);
        for (int k : key_list) {
            if (!visited[k]) {
                visited[k] = true;
                dfs(rooms, k);
            }
        }
    }

}
